### PR TITLE
Make sure that added R-groups have non-zero coordinates

### PR DESCRIPTION
**Pull request by ptosco**
Currently if cores have pre-defined 2D coordinates, e.g.
image

added R-groups do not look good as they have zero coordinates:
image

This PR fixes it:
image

After I had already written a fix for this using `MolOps::addHs()`, I realized that #3767 already had code to do the same in a more straightforward way by exposing `setHydrogenCoords()`.
So I thought that rather than creating more review work with yet another RGD-related PR I'd rewrite what I had already written using code from #3767.
I volunteer to fix conflicts in #3767 after this is merged (they are very straightforward to fix).

### CODE REVIEW

**Comment by bp-kelley** [comment]
This is good.  Part of me wonders if we should have an option or post-processing step to align the molecule to the drawn core.  This could make for easier visualizations.

**Comment by bp-kelley** [comment]
@ptosco if you fix the conflicts, I can recheck this.

**Comment by ptosco** [comment]
@bp-kelley Would you mind having a look at https://github.com/rdkit/rdkit/pull/3908 first, which I think is higher priority? Then I'll fix conflicts and merge this one. Thanks!

File: Code/GraphMol/molopstest.cpp [hunk 1]
```diff
@@ -8051,6 +8051,38 @@ void testRingFamilies() {
 void testRingFamilies() {}
 #endif
 
+void testSetTerminalAtomCoords() {
```
Comment (bp-kelley) [inline Code/GraphMol/molopstest.cpp:8054]: What happens when you add two terminal atoms to the same atom?

**Comment by ptosco** [comment]
@greglandrum @bp-kelley I have fixed the conflicts and added another test.
I have noticed that with current `master` something as innocent as this

```
import rdkit
from rdkit import Chem
from rdkit.Chem import rdRGroupDecomposition

core = Chem.MolFromSmiles("[1*]c1ccc([2*])cn1")
mol = Chem.MolFromSmiles("Fc1ccc([2*])cn1")
params = rdRGroupDecomposition.RGroupDecompositionParameters()
rgd = rdRGroupDecomposition.RGroupDecomposition(core, params)
rgd.Add(mol)
```

fails with

```
[22:29:48]

****
Pre-condition Violation
bad atom degree
Violation occurred on line 165 in file /scratch/toscopa1/src/rdkit/Code/GraphMol/AddHs.cpp
Failed Expression: mol->getAtomDegree(hydAtom) == 1
****

Traceback (most recent call last):
  File "/home/toscopa1/py/setHydrogenCoords.py", line 9, in <module>
    rgd.Add(mol)
RuntimeError: Pre-condition Violation
        bad atom degree
        Violation occurred on line 165 in file Code/GraphMol/AddHs.cpp
        Failed Expression: mol->getAtomDegree(hydAtom) == 1
        RDKIT: 2021.03.1dev1
        BOOST: 1_74
```

This PR also fixes this issue.
I have added a C++ test to cover the above case.
```

File: Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp [hunk 2]
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
...
+    TEST_ASSERT(rgdCore->getNumConformers() == 1);
+    size_t r2Num = 0;
+    for (const auto atom : rgdCore->atoms()) {
+      // test that R2 has non-zero coords and a sensible bond length
+      // to its neighboring atom
+      if (atom->getAtomicNum() == 0 && atom->getAtomMapNum() == 2) {
+        ++r2Num;
+        auto &r2Coord = rgdCore->getConformer().getAtomPos(atom->getIdx());
+        TEST_ASSERT(!(r2Coord.x > -1.e-4 && r2Coord.x < 1.e-4));
+        TEST_ASSERT(!(r2Coord.y > -1.e-4 && r2Coord.y < 1.e-4));
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:2177]:
```suggestion
        TEST_ASSERT(fabs(r2Coord.x)>1e-4);
        TEST_ASSERT(fabs(r2Coord.y)>1e-4);
```

File: Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp [hunk 3]
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
...
+        for (const auto &nbri :
+             boost::make_iterator_range(rgdCore->getAtomNeighbors(atom))) {
+          ++nBonds;
+          const auto nbr = (*rgdCore)[nbri];
+          const auto bond =
+              rgdCore->getBondBetweenAtoms(nbr->getIdx(), atom->getIdx());
+          TEST_ASSERT(bond);
+          auto &nbrCoord = rgdCore->getConformer().getAtomPos(nbr->getIdx());
+          auto bondLen = (nbrCoord - r2Coord).length();
+          TEST_ASSERT(bondLen > 0.9 && bondLen < 1.1);
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:2188]:
```suggestion
          TEST_ASSERT(fabs(bondLen - 1.0) < 0.1);
```

File: Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp [hunk 4]
```diff
@@ -2115,6 +2115,85 @@ void testSingleAtomBridge() {
...
+    size_t r2Num = 0;
+    for (const auto atom : rgdCore->atoms()) {
+      // test that R2 has non-zero coords and a sensible bond length
+      // to its neighboring atom
+      if (atom->getAtomicNum() == 0 && atom->getAtomMapNum() == 2) {
+        ++r2Num;
+        auto &r2Coord = rgdCore->getConformer().getAtomPos(atom->getIdx());
+        TEST_ASSERT(!(r2Coord.x > -1.e-4 && r2Coord.x < 1.e-4));
+        TEST_ASSERT(!(r2Coord.y > -1.e-4 && r2Coord.y < 1.e-4));
+        size_t nBonds = 0;
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/testRGroupDecomp.cpp:2178]: You don't need to count the number of bonds during the loop here.
The test you can do is very direct:
```
TEST_ASSERT(atom->getDegree()==1);

File: Code/GraphMol/MolOps.h [hunk 5]
```diff
@@ -192,6 +192,18 @@ RDKIT_GRAPHMOL_EXPORT void addHs(RWMol &mol, bool explicitOnly = false,
                                  const UINT_VECT *onlyOnAtoms = nullptr,
                                  bool addResidueInfo = false);
 
+//! Sets Cartesian coordinates for a terminal atom.
+//! Useful for growing an atom off a molecule with sensible
+//! coordinates based on the geometry of the neighbor.
+/*!
```
Comment (greglandrum) [inline Code/GraphMol/MolOps.h:198]:
```suggestion
//! coordinates based on the geometry of the neighbor.
/*!
    NOTE: this sets appropriate coordinates in all of the molecule's conformers.
```

File: Code/GraphMol/Wrap/MolOps.cpp [hunk 6]
```diff
@@ -1012,6 +1012,24 @@ struct molops_wrapper {
 \n";
     python::def("GetSymmSSSR", getSymmSSSR, docString.c_str());
 
+    // ------------------------------------------------------------------------
+    docString =
+        "Sets Cartesian coordinates for a terminal atom.\n\
+\n\
+  Useful for growing an atom off a molecule with sensible \n\
+  coordinates based on the geometry of the neighbor.\n\
+\n\
```
Comment (greglandrum) [inline Code/GraphMol/Wrap/MolOps.cpp:1021]:
```suggestion
  coordinates based on the geometry of the neighbor.\n\
\n\
  NOTE: this sets the appropriate coordinates in all of the molecule's conformers \n\
```

File: Code/GraphMol/molopstest.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/molopstest.cpp:8075]: Should be fine. The AddHs() code uses that all the time.

**Review by greglandrum — CHANGES_REQUESTED** [review]
small suggestions

**Comment by ptosco** [comment]
@greglandrum Thanks Greg, I have applied all your suggested changes, they all make sense.

**Review by greglandrum — APPROVED** [review]

**Comment by greglandrum** [comment]
I fixed the merge conflicts. Will merge after the CI builds pass

**Pull request merged**